### PR TITLE
loginctl: list-{users,sessions}: add a column for showing state

### PR DESCRIPTION
**Pull request by YHNdnzj**

### CODE REVIEW

**Review by yuwata — CHANGES_REQUESTED** [review]
LGTM.

Could you add a brief test for the fields in TEST-73-LOCALE ?

**Comment by dtardon** [comment]
@YHNdnzj I wonder: is the state really so useful to users that it needs to be listed? It seems nobody has ever asked for this...

**Comment by YHNdnzj** [comment]
> @YHNdnzj I wonder: is the state really so useful to users that it needs to be listed? It seems nobody has ever asked for this...

I personally encountered people asking for this :)

**Review by yuwata — APPROVED** [review]

**Comment by YHNdnzj** [comment]
@bluca I don't think the CI failures are related? `Lint Code Base` reports on code not changed in this PR, and `jammy-ppc64el` fails due to network errors.

**Comment by mrc0mmand** [comment]
> @bluca I don't think the CI failures are related? `Lint Code Base` reports on code not changed in this PR, and `jammy-ppc64el` fails due to network errors.

Don't worry about the shellcheck stuff - a new version of shellcheck got out with additional checks. I'll fix it when I do another round of shell-related fixes.

File: src/login/loginctl.c [hunk 1]
```diff
@@ -232,12 +233,25 @@ static int list_users(int argc, char *argv[], void *userdata) {
...
+                r = sd_bus_get_property_string(bus,
+                                               "org.freedesktop.login1",
+                                               object,
+                                               "org.freedesktop.login1.User",
+                                               "State",
+                                               &error,
+                                               &state);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to get state for user %s: %s",
+                                               user, bus_error_message(&error, r));
```
Comment (poettering) [inline src/login/loginctl.c]: this is race: by the time you query the state here the user might aleady have logged out fully. In which case you'll get SD_BUS_ERROR_UNKNOWN_OBJECT. I'd add a check for that and simply skip the entry in that case.

File: src/login/loginctl.c [hunk 2]
```diff
@@ -153,22 +153,16 @@ static int list_sessions(int argc, char *argv[], void *userdata) {
...
+                                               "TTY",
+                                               &error_tty,
+                                               &tty);
                 if (r < 0)
-                        log_warning_errno(r, "Failed to get TTY for session %s: %s", id, bus_error_message(&error_tty, r));
-                else {
-                        r = sd_bus_message_read(reply_tty, "s", &tty);
-                        if (r < 0)
-                                return bus_log_parse_error(r);
-                }
```
Comment (poettering) [inline src/login/loginctl.c:171]: maybe similar to my suggestion in the previous commit, catch SD_BUS_ERROR_UNKNOWN_OBJECT here and skip the entry

File: src/login/loginctl.c [hunk 3]
```diff
@@ -164,12 +164,24 @@ static int list_sessions(int argc, char *argv[], void *userdata) {
...
                                           id, bus_error_message(&error_tty, r));
 
+                r = sd_bus_get_property_string(bus,
+                                               "org.freedesktop.login1",
+                                               object,
+                                               "org.freedesktop.login1.Session",
+                                               "State",
+                                               &error,
+                                               &state);
+                if (r < 0)
```
Comment (poettering) [inline src/login/loginctl.c]: similar

**Pull request merged**